cppcache/src/PdxType.cpp

```diff
@@ -557,11 +557,30 @@ bool PdxType::Equals(std::shared_ptr<PdxType> otherObj) {
 }
 
 bool PdxType::operator<(const PdxType& other) const {
-  auto typeIdLessThan = this->m_geodeTypeId < other.m_geodeTypeId;
-  auto typeIdsBothZero =
-      (this->m_geodeTypeId == 0) && (other.m_geodeTypeId == 0);
-  auto classnameLessThan = this->m_className < other.m_className;
-  return (typeIdLessThan || (typeIdsBothZero && classnameLessThan));
+  if (m_geodeTypeId < other.m_geodeTypeId) {
+    return true;
+  }
+
+  if ((m_geodeTypeId == 0) && (other.m_geodeTypeId == 0)) {
+    return this->m_className < other.m_className;
+  }
+
+  return false;
+}
+
+int32_t PdxType::hashcode() const {
+  std::hash<std::string> strHash;
+  auto result=strHash(this->m_className);
+
+  for (std::vector<std::shared_ptr<PdxFieldType>>::iterator it =
+          m_pdxFieldTypes->begin();
+       it != m_pdxFieldTypes->end(); ++it) {
+    auto pdxPtr = *it;
+    result = result ^ (strHash(pdxPtr->getClassName()) << 1 );
```

Comment:
Do we need to xor in the classname each time through the loop?  It seems like just using the field names in here would be more appropriate

Reply:
One thing that is essential we get right here is that ordering of fields should have no bearing on the generated hash.  so { "foo": 7, "bar" 10 } is the same type as { "bar": 12, "foo": 11 }, e.g.  Otherwise we end up with a "type explosion" of zillions of copies of the same thing each getting their own type ID.  If we're lucky, the server side will sort the field names when generating the type ID on that side, and they'll all get the same ID, but then we've still made a ton of unnecessary trips to the server.

Reply:
> Do we need to xor in the classname each time through the loop? It seems like just using the field names in here would be more appropriate

I added it because I thought in the case of two different classes having the same fields (names & types).



> One thing that is essential we get right here is that ordering of fields should have no bearing on the generated hash. so { "foo": 7, "bar" 10 } is the same type as { "bar": 12, "foo": 11 }, e.g. Otherwise we end up with a "type explosion" of zillions of copies of the same thing each getting their own type ID. If we're lucky, the server side will sort the field names when generating the type ID on that side, and they'll all get the same ID, but then we've still made a ton of unnecessary trips to the server.

In the tests I did, I saw that the order of the fields does not change the result of the hash.

Reply:
Just to be clear, I didn't mean to suggest we not hash in the classname at all, just outside the for loop.  Also for what it's worth, even though using classname is still correct, the only instance of PDX we're aware of that has this problem, i.e. the classname isn't unique, is for JSON instances, where classname is always "__GEMFIRE_JSON".

Reply:
It would be good to have a unit test that shows the hashcode doesn't change if you iterate the fields in a different order...

Reply:
> Just to be clear, I didn't mean to suggest we not hash in the classname at all, just outside the for loop. Also for what it's worth, even though using classname is still correct, the only instance of PDX we're aware of that has this problem, i.e. the classname isn't unique, is for JSON instances, where classname is always "__GEMFIRE_JSON".

oh, I see the confusion: the classname its outside the for loop (line 573). The other classnames you are refering to are the classnames of each field. If we dont include the name of the field and its classname, the hash will be the same for an object with a bool called `foo` and for an object with a string called `foo`.

Reply:
> It would be good to have a unit test that shows the hashcode doesn't change if you iterate the fields in a different order...

sure, Im working on that :)

Reply:
> > Just to be clear, I didn't mean to suggest we not hash in the classname at all, just outside the for loop. Also for what it's worth, even though using classname is still correct, the only instance of PDX we're aware of that has this problem, i.e. the classname isn't unique, is for JSON instances, where classname is always "__GEMFIRE_JSON".
> 
> oh, I see the confusion: the classname its outside the for loop (line 573). The other classnames you are refering to are the classnames of each field. If we dont include the name of the field and its classname, the hash will be the same for an object with a bool called `foo` and for an object with a string called `foo`.

I don't think the Java implementation will distinguish one OBJECT from anther OBJECT . I think you just need to has the field name and type and ignore the class name.